File: services/device/src/main/java/com/example/iter/device/controller/api/EquipmentQueryApiController.java
package com.example.iter.device.controller.api;

import com.example.iter.common.security.CustomUserDetails;
import com.example.iter.device.controller.api.spec.EquipmentQueryApiSpec;
import com.example.iter.device.dto.request.EquipmentAvailabilityRequest;
import com.example.iter.device.dto.request.EquipmentEstimateRequest;
import com.example.iter.device.dto.request.EquipmentSearchRequest;
import com.example.iter.device.dto.response.EquipmentAvailabilityResponse;
import com.example.iter.device.dto.response.EquipmentDetailResponse;
import com.example.iter.device.dto.response.EquipmentEstimateResponse;
import com.example.iter.device.dto.response.EquipmentListResponse;
import com.example.iter.device.service.EquipmentQueryService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/devices")
@RequiredArgsConstructor
public class EquipmentQueryApiController implements EquipmentQueryApiSpec {

    private final EquipmentQueryService equipmentQueryService;

    @Override
    @GetMapping("/{equipmentId}/availability")
    public ResponseEntity<EquipmentAvailabilityResponse> getEquipmentAvailability(
            @PathVariable Long equipmentId,
            @Valid @ModelAttribute EquipmentAvailabilityRequest request
    ) {
        return ResponseEntity.ok(
                equipmentQueryService.getEquipmentAvailability(equipmentId, request));
    }

    @Override
    @GetMapping("/{equipmentId}/estimate")
    public ResponseEntity<EquipmentEstimateResponse> getEquipmentEstimate(
            @PathVariable Long equipmentId,
            @Valid @ModelAttribute EquipmentEstimateRequest request
    ) {
        return ResponseEntity.ok(equipmentQueryService.getEquipmentEstimate(equipmentId, request));
    }

    @Override
    @GetMapping("/{equipmentId}")
    public ResponseEntity<EquipmentDetailResponse> getEquipmentDetail(
            @AuthenticationPrincipal CustomUserDetails principal,
            @PathVariable Long equipmentId
    ) {
        Long requesterId = principal == null ? null : principal.getUser().getId();
        return ResponseEntity.ok(equipmentQueryService.getEquipmentDetail(requesterId, equipmentId));
    }

    @Override
    @GetMapping
    public ResponseEntity<EquipmentListResponse> getEquipmentList(
            @Valid @ModelAttribute EquipmentSearchRequest request
    ) {
        return ResponseEntity.ok(equipmentQueryService.getEquipmentList(request));
    }
}

File: services/device/src/main/kotlin/com/example/iter/device/controller/api/EquipmentQueryApiController.kt
package com.example.iter.device.controller.api

import com.example.iter.common.security.CustomUserDetails
import com.example.iter.device.controller.api.spec.EquipmentQueryApiSpec
import com.example.iter.device.dto.request.EquipmentAvailabilityRequest
import com.example.iter.device.dto.request.EquipmentEstimateRequest
import com.example.iter.device.dto.request.EquipmentSearchRequest
import com.example.iter.device.dto.response.EquipmentAvailabilityResponse
import com.example.iter.device.dto.response.EquipmentDetailResponse
import com.example.iter.device.dto.response.EquipmentEstimateResponse
import com.example.iter.device.dto.response.EquipmentListResponse
import com.example.iter.device.service.EquipmentQueryService
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/devices")
class EquipmentQueryApiController(
    private val equipmentQueryService: EquipmentQueryService,
) : EquipmentQueryApiSpec {

    @GetMapping("/{equipmentId}/availability")
    override fun getEquipmentAvailability(
        @PathVariable("equipmentId") equipmentId: Long,
        @Valid @ModelAttribute request: EquipmentAvailabilityRequest,
    ): ResponseEntity<EquipmentAvailabilityResponse> =
        ResponseEntity.ok(equipmentQueryService.getEquipmentAvailability(equipmentId, request))

    @GetMapping("/{equipmentId}/estimate")
    override fun getEquipmentEstimate(
        @PathVariable("equipmentId") equipmentId: Long,
        @Valid @ModelAttribute request: EquipmentEstimateRequest,
    ): ResponseEntity<EquipmentEstimateResponse> =
        ResponseEntity.ok(equipmentQueryService.getEquipmentEstimate(equipmentId, request))

    @GetMapping("/{equipmentId}")
    override fun getEquipmentDetail(
        @AuthenticationPrincipal principal: CustomUserDetails?,
        @PathVariable("equipmentId") equipmentId: Long,
    ): ResponseEntity<EquipmentDetailResponse> {
        val requesterId = principal?.user?.id
        return ResponseEntity.ok(equipmentQueryService.getEquipmentDetail(requesterId, equipmentId))
    }

    @GetMapping
    override fun getEquipmentList(
        @Valid @ModelAttribute request: EquipmentSearchRequest,
    ): ResponseEntity<EquipmentListResponse> =
        ResponseEntity.ok(equipmentQueryService.getEquipmentList(request))
}
